### PR TITLE
OnlyOrg Policy

### PR DESCRIPTION
**Pull request by addisonbeck**
## Narrative  ✍🏻
As an enterprise organization admin I want to be able to block members of my organization from being able to be a part of other organizations.

## Acceptance Criteria 💯
1. A policy (OnlyOrg) exists that restricts non-admin members from being able to be a part of other organizations.
2. When applying the OnlyOrg policy any non-owner members who aren't the one enabling the policy & that are already in another org should be removed from the org applying the policy. This user flow should mirror what already exists for the 2FA policy.
3. Admins should be warned about users potentially being removed before enabling the Only Org policy
4. Removed users should get an email informing them of the removal.
5. The OnlyOrg policy needs to be configurable from the web vault AND the enterprise portal.
6. A message is displayed on the web vault's policy page warning users that it will soon be deprecated and they should use the enterprise portal instead.
7. Users in an org with the OnlyOrg policy enabled should not be able to create new organizations.

## Code Changes 👨🏻‍💻

- [x] added any needed localization strings for the feature
- [x] added a deprecation warning to the policies page using a clickable callout that redirects to enterprise
- [x] allowed onlyOrg policy configuration, with warning about potential user kicks
- [x] blocked creating new orgs if already in an org with OnlyOrg enabled by cancelling the submit function and throwing up a callout above the button

## Screenshots
Screen Shot 2020-10-08 at 1 24 43 PM 1
Screen Shot 2020-10-08 at 1 24 50 PM
Screen Shot 2020-10-08 at 1 25 09 PM

## Notes 📝
[Asana ticket](https://app.asana.com/0/1195018406457675/1196488889925152)
[Server PR](https://github.com/bitwarden/server/pull/962)
[jslib PR](https://github.com/bitwarden/jslib/pull/183)

*jslib will need to be updated to finalize this PR

### CODE REVIEW

File: src/app/organizations/manage/policies.component.html [hunk 1]
```diff
@@ -1,3 +1,6 @@
+<button class="p-0 deprecationWarning" (click)="goToEnterprisePortal()">
+    <app-callout [type]="'warning'">{{'webPoliciesDeprecationWarning' | i18n}}</app-callout>
+</button>
```
Comment (kspearrin) [inline src/app/organizations/manage/policies.component.html:3]: Can we follow this style for callout action buttons?

image

File: src/app/settings/organization-plans.component.ts [hunk 2]
```diff
@@ -28,6 +29,7 @@ import { ProductType } from 'jslib/enums/productType';
 import { OrganizationCreateRequest } from 'jslib/models/request/organizationCreateRequest';
 import { OrganizationUpgradeRequest } from 'jslib/models/request/organizationUpgradeRequest';
 import { PlanResponse } from 'jslib/models/response/planResponse';
+import { PolicyType } from 'jslib/enums/policyType';
```
Comment (kspearrin) [inline src/app/settings/organization-plans.component.ts:32]: group it with existing enums

File: src/app/settings/organization-plans.component.ts [hunk 3]
```diff
@@ -204,6 +208,17 @@ export class OrganizationPlansComponent implements OnInit {
...
+        if (this.onlyOrgPolicyBlock) {
+            return;
+        }
+        else {
+            const policies = await this.policyService.getAll(PolicyType.OnlyOrg);
+            this.onlyOrgPolicyBlock = policies.filter(policy => policy.enabled).length > 0;
+            if (this.onlyOrgPolicyBlock) {
+                return;
+            }
+        }
```
Comment (kspearrin) [inline src/app/settings/organization-plans.component.ts:220]: formatting on if/else. Might want to do this first, before the file parsing above.

File: src/locales/en/messages.json [hunk 4]
```diff
@@ -3206,5 +3206,20 @@
...
+    "message": "Policy configuration has been moved, and this page will soon be deprecated. Please click here to use the Enterprise Portal policies page instead."
+  },
+  "onlyOrg": {
+    "message": "Only Organization"
+  },
+  "onlyOrgDesc": {
+    "message": "Restrict users from being able to join any other organizations."
+  },
+  "onlyOrgBlockCreateMessage": {
+    "message": "You may not perform this action because your current organization restricts users from having multiple. Please contact your organization admins or sign up from a different Bitwarden account."
```
Comment (kspearrin) [inline src/locales/en/messages.json:3220]: > Your current organization has a policy that does not allow you to join more than one organization. Please contact your organization admins or sign up from a different Bitwarden account.

File: src/locales/en/messages.json [hunk 5]
```diff
@@ -3206,5 +3206,20 @@
...
+  },
+  "onlyOrgPolicyWarning": {
+    "message": "Organization members who are already a part of another organization will be removed from this organization and will receive an email notifying them about the change."
```
Comment (kspearrin) [inline src/locales/en/messages.json:3223]: > Organization members who are already a member of another organization will be removed from your organization.

File: src/scss/styles.scss [hunk 6]
```diff
@@ -659,13 +659,15 @@ app-user-billing {
     }
 }
 
+$callout-bg: #fafafa;
```
Comment (kspearrin) [inline src/scss/styles.scss:662]: I'd prefer that we move to a standard button styling like the above screenshot and revert this.

**Review by kspearrin — CHANGES_REQUESTED** [review]

File: src/app/organizations/manage/policies.component.html (hunk 1 above)
Comment (addisonbeck) [reply src/app/organizations/manage/policies.component.html]: done

File: src/app/settings/organization-plans.component.ts (hunk 2 above)
Comment (addisonbeck) [reply src/app/settings/organization-plans.component.ts]: done

File: src/app/settings/organization-plans.component.ts (hunk 3 above)
Comment (addisonbeck) [reply src/app/settings/organization-plans.component.ts]: fixed + moved

File: src/locales/en/messages.json (hunk 4 above)
Comment (addisonbeck) [reply src/locales/en/messages.json]: 👍

File: src/locales/en/messages.json (hunk 5 above)
Comment (addisonbeck) [reply src/locales/en/messages.json]: 👍

File: src/scss/styles.scss (hunk 6 above)
Comment (addisonbeck) [reply src/scss/styles.scss]: removed

File: src/app/settings/organization-plans.component.ts [hunk 7]
```diff
@@ -197,6 +197,16 @@ export class OrganizationPlansComponent implements OnInit {
     }
 
     async submit() {
+        if (this.onlyOrgPolicyBlock) {
+            return;
+        } else {
+            const policies = await this.policyService.getAll(PolicyType.OnlyOrg);
+            this.onlyOrgPolicyBlock = policies.filter(policy => policy.enabled).length > 0;
```
Comment (kspearrin) [inline src/app/settings/organization-plans.component.ts:204]: I think you can use `policies.some(policy => policy.enabled)` here to be more efficient.

**Review by kspearrin — CHANGES_REQUESTED** [review]

File: src/app/settings/organization-plans.component.ts (hunk 7 above)
Comment (addisonbeck) [reply src/app/settings/organization-plans.component.ts]: fixed

**Review by kspearrin — APPROVED** [review]

**Pull request merged**